configure.ac

```diff
@@ -0,0 +1,1666 @@
+    [use frame pointers in runtime and generated code])])
+
+AC_ARG_ENABLE([naked-pointers],
+  [AS_HELP_STRING([--disable-naked-pointers],
+    [do not allow naked pointers])])
+
+AC_ARG_ENABLE([spacetime],
+  [AS_HELP_STRING([--enable-spacetime],
+    [build the spacetime profiler])])
+
+AC_ARG_ENABLE([call-counts],
+  [AS_HELP_STRING([--disable-call-counts],
+    [disable the call counts in spacetime])])
+
+AC_ARG_ENABLE([cfi],
+  [AS_HELP_STRING([--disable-cfi],
+    [disable the CFI directives in assembly files])])
+
+AC_ARG_ENABLE([installing-source-artifacts],
+  [AS_HELP_STRING([--enable-installing-source-artifacts],
+    [install *.cmt* and *.mli files])])
+AC_ARG_ENABLE([installing-bytecode-programs],
+  [AS_HELP_STRING([--enable-installing-bytecode-programs],
+    [also install the bytecode versions of programs])])
+
+AC_ARG_ENABLE([native-compiler],
+  [AS_HELP_STRING([--disable-native-compiler],
+    [do not build the native compiler])])
+
+AC_ARG_ENABLE([flambda],
+  [AS_HELP_STRING([--enable-flambda],
+    [enable flmabda optimizations])])
+
+AC_ARG_ENABLE([flambda-invariants],
+  [AS_HELP_STRING([--enable-flambda-invariants],
+    [enable invariants checks in flambda])])
+
+AC_ARG_WITH([target-bindir],
+  [AS_HELP_STRING([--with-target-bindir],
+    [location of binary programs on target system])])
+
+AC_ARG_WITH([reserved-header-bits],
+  [AS_HELP_STRING([--with-reserved-header-bits=BITS],
+    [reserve BITS bits in block headers for profiling info])])
+
+# There are two configure-time string safety options,
+# --(enable|disable)-force-safe-string and
+# --with-default-string=safe|unsafe that
+# interact with a compile-time (un)safe-string option.
+#
+# If --enable-force-safe-string is set at configure time, then the compiler
+# will always enforce that string and bytes are distinct: the
+# compile-time -unsafe-string option is disabled. This lets us
+# assume pervasive string immutability, for code optimizations and
+# in the C layer.
+#
+# If --disable-force-safe-string is set at configure-time, the compiler
+# will use the compile-time (un)safe-string option to decide whether
+# string and bytes are compatible on a per-file basis. The
+# configure-time options --with-default-string=safe|unsafe decides which
+# setting will be chosen by default, if no compile-time option is
+# explicitly passed.
+#
+# The configure-time behavior of OCaml 4.05 and older was equivalent
+# to --disable-force-safe-string --with-default-string=unsafe. OCaml 4.06
+# and later uses --disable-force-safe-string --with-default-string=safe. We
+# expect --enable-force-safe-string to become the default in the future.
+
+AC_ARG_ENABLE([force-safe-string],
+  [AS_HELP_STRING([--disable-force-safe-string],
+    [do not force strings to be safe])])
+
+AC_ARG_WITH([default-string],
+  [AS_HELP_STRING([--with-default-string=mode],
+    [whether strings should be safe or unsafe by default
+      @<:@default=safe@:>@])])
```

Comment:
In trunk, `-force-safe-string` takes precedence. (see lines 167-170 of `utils/clflags.ml`)

Reply:
I'd use `unsafe / safe / force` instead of `no / yes / always`.

Reply:
There is a conceptual difference between:
- `--force-safe-string`, which must be a configure-time option that changes the runtime definitions and has a global impact on all modules compiled and linked by the resulting compiler
- `--default-string={safe,unsafe}` which is "merely" a convenience feature to set a default for compile-time options, rather than a configure-time option by nature (`afl` is the same, etc.). It does nothing more than change the default for a choice that the user makes on a per-invocation basis.

Reply:
In my opinion this conceptual difference justifies *not* showing the three options as a single knob to fix at configure-time, but two separate kind of options. There is a dependency between then (`--force-safe-string` makes it incorrect to change the per-invocation value, so it makes `--default-string` useless and meaningless), but that is common across the configuration space, for example `--disable-native-compiler` will make many other settings meaningless.